### PR TITLE
feat: Include partitioned tables into the schema cache

### LINKED ISSUE #1783: Partitioned tables are not supported by everything related to schema cache
### Environment

* PostgreSQL version: with docker: `postgres:13.2`
* PostgREST version: with docker: `postgrest/postgrest:nightly-2021-03-05-19-03-d3a8b5f`
* Operating system: MacOS

### Description of issue
This issue is related #1592 (and the fix: #1593)
I've tested the last the nightly builds and I still have the same issue, but **only with partitioned tables**.

It seem that PostgREST don't recognize foreign keys if the tables are partitioned, but if I remove the partitions all work fine.

```SQL
CREATE TABLE public."Diary"
(
  school text NOT NULL,
  day text NOT NULL,
  student text NOT NULL,
  rif text NOT NULL GENERATED ALWAYS AS (immutable_concat_ws('|',school,day,student)) STORED,
  room text NOT NULL,
  -- ...other...
  PRIMARY KEY (school, day, student),
  FOREIGN KEY (school) REFERENCES public."School" (id),
  FOREIGN KEY (school, student) REFERENCES public."Student" (school, student),
  FOREIGN KEY (school, room) REFERENCES public."Room" (school, room),
  CONSTRAINT CHECK_Diary_day CHECK (day ~ '^\d{4}\-(0[1-9]|1[012])\-(0[1-9]|[12][0-9]|3[01])$')
)  PARTITION BY LIST (school);
-- ... (indexes)

CREATE TABLE public."Student"
(
  school text NOT NULL,
  student text NOT NULL,
  rif text NOT NULL GENERATED ALWAYS AS (school || '|' || student) STORED,
  room text NOT NULL,
  -- ...other...
  PRIMARY KEY (school, student),
  FOREIGN KEY (school) REFERENCES public."School" (id),
  FOREIGN KEY (school, room) REFERENCES public."Room" (school, room)
) PARTITION BY LIST (school);
-- ... (indexes)

```

### PR DESCRIPTION
**Pull request by laurenceisla**
Closes #1783

Tests:

- [x] Embedding
- [x] Inserts
- [x] Upserts
- [x] Options
- [x] OpenAPI

### CODE REVIEW

File: test/fixtures/schema.sql [hunk 1]
```diff
@@ -2185,3 +2185,61 @@ create table private.rollen (
   foreign key (film_id) references test.filme(id),
   foreign key (rolle_id) references test.schauspieler(id)
 );
+
+-- Tables used for testing embedding between partitioned tables
+
+DO $do$BEGIN
+    -- partitioned tables using the PARTITION syntax are supported from pg v10
+    IF (SELECT current_setting('server_version_num')::INT >= 100000) THEN
```
Comment (steve-chavez) [inline test/fixtures/schema.sql:2193]: Wow, this feature is really complete!

Let's keep all the identifiers as lowercase for consistency:

```suggestion
    -- partitioned tables using the PARTITION syntax are supported from pg v10
    if (select current_setting('server_version_num')::int >= 100000) then
```

File: test/fixtures/schema.sql [hunk 2]
```diff
@@ -2185,3 +2185,61 @@ create table private.rollen (
...
+
+      create table test.first_partition_a partition of test.partitioned_a
+        for values in ('first');
+
+      create table test.second_partition_a partition of test.partitioned_a
+        for values in ('second');
+    END IF;
+
+    -- primary keys for partitioned tables are supported from pg v11
+    IF (SELECT current_setting('server_version_num')::INT >= 110000) THEN
```
Comment (steve-chavez) [inline test/fixtures/schema.sql:2212]: Ditto.
```suggestion
    -- primary keys for partitioned tables are supported from pg v11
    if (select current_setting('server_version_num')::int >= 110000) then
```

File: test/fixtures/schema.sql [hunk 3]
```diff
@@ -2185,3 +2185,61 @@ create table private.rollen (
...
+      create table test.reference_from_partitioned (
+        id int primary key
+      );
+
+      alter table test.partitioned_a add primary key (id, name);
+      alter table test.partitioned_a add column id_ref int references test.reference_from_partitioned(id);
+    END IF;
+
+    -- foreign keys referencing partitioned tables are supported from pg v12
+    IF (SELECT current_setting('server_version_num')::INT >= 120000) THEN
```
Comment (steve-chavez) [inline test/fixtures/schema.sql:2222]: Ditto.
```suggestion
    -- foreign keys referencing partitioned tables are supported from pg v12
    if (select current_setting('server_version_num')::int >= 120000) then
```

File: test/Feature/UpsertSpec.hs [hunk 4]
```diff
@@ -43,6 +45,20 @@ spec =
...
 
+        when (actualPgVersion >= pgVersion110) $
+          it "INSERTs and UPDATEs rows on composite pk conflict for partitioned tables" $
+            request methodPost "/partitioned_a" [("Prefer", "return=representation"), ("Prefer", "resolution=merge-duplicates")]
+              [json| [
+                { "id": 4, "name": "second", "id_ref": 2},
+                { "id": 6, "name": "first", "id_ref": 1 }
+              ]|] `shouldRespondWith` [json| [
+                { "id": 4, "name": "second", "id_ref": 2 },
+                { "id": 6, "name": "first", "id_ref": 1 }
```
Comment (steve-chavez) [inline test/Feature/UpsertSpec.hs:56]: I see the `id: 6` value is inserted here, but is the `id: 4` updated in any column? I see it already has "second" in the fixtures.

File: test/Feature/QuerySpec.hs [hunk 5]
```diff
@@ -395,6 +396,76 @@ spec actualPgVersion = do
...
+                 {"id":4,"name":"second_b","second_partition_a":{"id":4,"name":"second"}}] |]
+              { matchHeaders = [matchContentTypeJson] }
+
+            get "/second_partition_a?select=id,name,partitioned_b(id,name)&order=id.asc" `shouldRespondWith`
+              [json|
+                [{"id":3,"name":"second","partitioned_b":[]},
+                 {"id":4,"name":"second","partitioned_b":[{"id":4,"name":"second_b"}]}] |]
+              { matchHeaders = [matchContentTypeJson] }
+
+          it "can request parents and children between a table and a partitioned table" $ do
```
Comment (steve-chavez) [inline test/Feature/QuerySpec.hs:456]: From a glance, this looks the same as the above test that says:

```haskell
it "can request parents and children between a partitioned table and a table"
```

But the difference is in the target of the foreign key right? Perhaps you can clarify that on the specs.

File: src/PostgREST/DbStructure.hs [hunk 6]
```diff
@@ -323,46 +323,55 @@ accessibleTables =
...
-        AND (pg_relation_is_updatable(c.oid::regclass, FALSE) & 8) = 8
-        OR EXISTS (
-          SELECT 1
-          FROM pg_trigger
-          WHERE
-            pg_trigger.tgrelid = c.oid
-            AND (pg_trigger.tgtype::integer & 69) = 69
+        c.relkind IN ('r','p')
+        OR (
+          c.relkind in ('v','f')
```
Comment (steve-chavez) [inline src/PostgREST/DbStructure.hs:328]: Did you run this new query on the big schema? Seems the perf won't change, but maybe we should double check to be sure.

**Comment by steve-chavez** [comment]
This is great work @laurenceisla.

I've noted that this is in fact a feature and not a fix, so it should be documented in a subsection on [Resource Embedding](https://postgrest.org/en/v7.0.0/api.html#resource-embedding). Maybe there it can be mentioned that embedding on partitions is fully supported on pg 12.

**Comment by laurenceisla** [comment]
> I've noted that this is in fact a feature and not a fix, so it should be documented in a subsection on [Resource Embedding](https://postgrest.org/en/v7.0.0/api.html#resource-embedding). Maybe there it can be mentioned that embedding on partitions is fully supported on pg 12.

@steve-chavez Yes, I also think it should be documented at least on the embedding section, due to the issue opened is related to that.

> Great work! By the way, the two check failures should go away with a rebase on main:

@monacoremo  Cool! I'll merge to fix the CI failures.

File: test/Feature/UpsertSpec.hs (hunk 4 above)
Comment (laurenceisla) [reply test/Feature/UpsertSpec.hs:56]: Yes, it's a composite primary key `(id, name)`. The column that is updated is `id_ref`, from `null` to `2`.

Comment (steve-chavez) [reply test/Feature/UpsertSpec.hs:56]: I see, then it's good to leave it as it is :+1:

File: src/PostgREST/DbStructure.hs (hunk 6 above)
Comment (laurenceisla) [reply src/PostgREST/DbStructure.hs]: Just tested it with the big schema and couldn't find relevant differences in the execution time between this branch and main.

**Review by steve-chavez — APPROVED** [review]

**Comment by monacoremo** [comment]
We should also add a changelog entry!

Not sure how the merge you did will interact when merging the pull request... would be good to rebase on main and remove the merge commit!

**Comment by laurenceisla** [comment]
> We should also add a changelog entry!

Done.

> Not sure how the merge you did will interact when merging the pull request... would be good to rebase on main and remove the merge commit!

I think a squash and merge should handle that conflict, it'll look as only one commit.

**Comment by wolfgangwalther** [comment]
> Codecov: We introduced a 1% threshold to silence the random coverage fluctuations and resulting failures - for now, until we figure out their cause

Cool, I like that.

**Comment by BakerHughesCharlesSewell** [comment]
I noticed that this pull request is in the Upcoming Release backlog. Is there a nightly build containing these code changes? It looks like the last nightly build was on July 7th.

**Comment by steve-chavez** [comment]
@BakerHughesCharlesSewell We're not doing nightlies anymore but you can get the binaries built for this commit on the artifacts part(at the bottom) here: https://github.com/PostgREST/postgrest/actions/runs/1128881855

**Comment by steve-chavez** [comment]
@BakerHughesCharlesSewell We just made a pre-release with this change, you can download the latest binaries there: https://github.com/PostgREST/postgrest/releases/tag/v9.0.0-a1

**Comment by BakerHughesCharlesSewell** [comment]
I tested the following scenarios using the Film/Actor/Director example provided on the PostgREST documentation website under [Resource Embedded section](https://postgrest.org/en/v8.0/api.html#resource-embedding). Here is  SQL script to create both partitioned and non-partitioned tables.


image

**Pull request merged**